File: bench/benches/bell.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "lib"))
from bench import bench

# Bell numbers in base 36 (see bench/benches/bell.g). Fresh memo dicts per rep.
# checksum = total characters across all rendered lines.
DIGITS = "0123456789abcdefghijklmnopqrstuvwxyz"
BASE = len(DIGITS)

def bell_run(limit):
    facts, bells = {}, {}

    def fact(n):
        if n in facts:
            return facts[n]
        x, m = 1, n
        while m > 1:
            x *= m
            m -= 1
        facts[n] = x
        return x

    def choose(n, k):
        return fact(n) // (fact(k) * fact(n - k))

    def bell(n):
        if n in bells:
            return bells[n]
        r = 1 if n < 2 else sum(choose(n - 1, k) * bell(k) for k in range(n))
        bells[n] = r
        return r

    def show(n):
        s = ""
        while n > 0:
            s = DIGITS[n % BASE] + s
            n //= BASE
        return s

    total, i = 0, 0
    while True:
        b = show(bell(i))
        if len(b) > limit:
            return total
        total += len(b)
        i += 1

bench("bell", lambda: bell_run(280))

File: bench/benches/bell.jl
include(joinpath(@__DIR__, "..", "lib", "bench.jl"))

# Bell numbers in base 36 (see bench/benches/bell.l). Fresh memo dicts per rep.
# checksum = total characters across all rendered lines. Needs bignums (BigInt).
const BELL_DIGITS = "0123456789abcdefghijklmnopqrstuvwxyz"

function bell_run(limit)
    facts = Dict{Int,BigInt}()
    bells = Dict{Int,BigInt}()

    function fact(n)
        haskey(facts, n) && return facts[n]
        x = BigInt(1)
        m = n
        while m > 1
            x *= m
            m -= 1
        end
        facts[n] = x
        return x
    end

    choose(n, k) = fact(n) ÷ (fact(k) * fact(n - k))

    function bell(n)
        haskey(bells, n) && return bells[n]
        r = n < 2 ? BigInt(1) : sum(choose(n - 1, k) * bell(k) for k in 0:n-1)
        bells[n] = r
        return r
    end

    function show36(n)
        s = ""
        while n > 0
            s = string(BELL_DIGITS[Int(n % 36)+1]) * s
            n = n ÷ 36
        end
        return s
    end

    total = 0
    i = 0
    while true
        b = show36(bell(i))
        if length(b) > limit
            return total
        end
        total += length(b)
        i += 1
    end
end

bench("bell", () -> bell_run(280))
